Added functional benchmarks

archisgore opened the pull request:
This change adds two types of tests:
1. First some regular Test* tests that functionally
ensure the test cases are correct.

2. A few Benchmark* tests that then allow us to
measure Otto's performance under various functional scenarios.

The basic tests are just sorting, but I'm planning over time
to bring in more advanced tests (like the CryptoAES) from
popular benchmarking suites (JetStream for instance.)

I intend to run profiler on these tests to find any
ottlenecks or low-hanging fruit that can help speed up
Otto for real-world use-cases.

Next up after this is to add some heavy underscorejs
benchmarks.

stevenh commented on functional_benchmark_test.go:
```diff
@@ -0,0 +1,727 @@
+func BenchmarkCryptoAES(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		vm := New()
+		vm.Run(jsCryptoAES)
```
A not insignificant amount of the time could be taken to create the vm here

stevenh commented on functional_benchmark_test.go:
```diff
@@ -0,0 +1,727 @@
+func benchmarkJsArraySort(size int, sortFuncCall string, sortCode string, b *testing.B) {
+	// generate arbitrary slice of 'size'
+	testSlice := make([]string, size)
+	for i := 0; i < size; i++ {
```
Range over the slice

stevenh commented on functional_benchmark_test.go:
```diff
@@ -0,0 +1,727 @@
+    pivot = right;
+    partitionIndex = partition(arr, pivot, left, right);
+
+   //sort left and right
```
Style space after if and before { more in the rest of the js

stevenh replied:
Also space after //

The comment about style refers to a few lines up but can't edit on a mobile 😋

stevenh commented on functional_benchmark_test.go:
```diff
@@ -0,0 +1,727 @@
+
+if (decryptedText != plainText)
+    throw "ERROR: bad result: expected " + plainText + " but got " + decryptedText;
+`
```
New line at the end of the file

archisgore replied:
Done.

archisgore commented:
1. Moved vm creation outside of BenchmarchCryptoAES loop. Reset timer after it is created.

2. Added a space after all "//"

3. Added newline at the end of the file.

archisgore commented:
(This is purely for my reference so I can access it later)

BenchmarkNativeCallWithString-8                     	  500000	      2288 ns/op	     560 B/op	      10 allocs/op
BenchmarkNativeCallWithFloat32-8                    	  500000	      2533 ns/op	     568 B/op	      12 allocs/op
BenchmarkNativeCallWithFloat64-8                    	 1000000	      2268 ns/op	     560 B/op	      10 allocs/op
BenchmarkNativeCallWithInt-8                        	  500000	      2542 ns/op	     576 B/op	      12 allocs/op
BenchmarkNativeCallWithUint-8                       	  500000	      2525 ns/op	     576 B/op	      12 allocs/op
BenchmarkNativeCallWithInt8-8                       	  500000	      2527 ns/op	     562 B/op	      12 allocs/op
BenchmarkNativeCallWithUint8-8                      	  500000	      2516 ns/op	     562 B/op	      12 allocs/op
BenchmarkNativeCallWithInt16-8                      	  500000	      2536 ns/op	     564 B/op	      12 allocs/op
BenchmarkNativeCallWithUint16-8                     	  500000	      2550 ns/op	     564 B/op	      12 allocs/op
BenchmarkNativeCallWithInt32-8                      	  500000	      2552 ns/op	     568 B/op	      12 allocs/op
BenchmarkNativeCallWithUint32-8                     	  500000	      2539 ns/op	     568 B/op	      12 allocs/op
BenchmarkNativeCallWithInt64-8                      	 1000000	      2272 ns/op	     560 B/op	      10 allocs/op
BenchmarkNativeCallWithUint64-8                     	  500000	      2537 ns/op	     576 B/op	      12 allocs/op
BenchmarkNativeCallWithStringInt-8                  	  500000	      2875 ns/op	     640 B/op	      13 allocs/op
BenchmarkNativeCallWithIntVariadic0-8               	 1000000	      2052 ns/op	     560 B/op	      10 allocs/op
BenchmarkNativeCallWithIntVariadic1-8               	  300000	      5199 ns/op	    1177 B/op	      26 allocs/op
BenchmarkNativeCallWithIntVariadic3-8               	  300000	      3723 ns/op	     896 B/op	      21 allocs/op
BenchmarkNativeCallWithIntVariadic10-8              	  200000	      5869 ns/op	    1793 B/op	      37 allocs/op
BenchmarkNativeCallWithIntArray0-8                  	  500000	      3513 ns/op	    1172 B/op	      18 allocs/op
BenchmarkNativeCallWithIntArray1-8                  	  300000	      5181 ns/op	    1337 B/op	      28 allocs/op
BenchmarkNativeCallWithIntArray3-8                  	  200000	      7330 ns/op	    1745 B/op	      45 allocs/op
BenchmarkNativeCallWithIntArray10-8                 	  100000	     15853 ns/op	    4114 B/op	      99 allocs/op
BenchmarkNativeCallWithIntVariadicArray0-8          	  500000	      3694 ns/op	    1268 B/op	      20 allocs/op
BenchmarkNativeCallWithIntVariadicArray1-8          	  200000	      5467 ns/op	    1433 B/op	      30 allocs/op
BenchmarkNativeCallWithIntVariadicArray3-8          	  200000	      7872 ns/op	    1841 B/op	      47 allocs/op
BenchmarkNativeCallWithIntVariadicArray10-8         	  100000	     15839 ns/op	    4210 B/op	     101 allocs/op
BenchmarkNativeCallWithStringIntVariadic0-8         	  500000	      2570 ns/op	     640 B/op	      12 allocs/op
BenchmarkNativeCallWithStringIntVariadic1-8         	  300000	      5656 ns/op	    1257 B/op	      27 allocs/op
BenchmarkNativeCallWithStringIntVariadic3-8         	  500000	      4136 ns/op	     928 B/op	      21 allocs/op
BenchmarkNativeCallWithStringIntVariadic10-8        	  200000	      6164 ns/op	    1857 B/op	      37 allocs/op
BenchmarkNativeCallWithStringIntVariadicArray0-8    	  300000	      4115 ns/op	    1333 B/op	      21 allocs/op
BenchmarkNativeCallWithStringIntVariadicArray1-8    	  200000	      5849 ns/op	    1497 B/op	      31 allocs/op
BenchmarkNativeCallWithStringIntVariadicArray3-8    	  200000	      7865 ns/op	    1905 B/op	      48 allocs/op
BenchmarkNativeCallWithStringIntVariadicArray10-8   	  100000	     16488 ns/op	    4274 B/op	     102 allocs/op
BenchmarkNativeCallWithMap-8                        	  300000	      4698 ns/op	    1665 B/op	      24 allocs/op
BenchmarkNativeCallWithMapVariadic-8                	  200000	      7204 ns/op	    2913 B/op	      42 allocs/op
BenchmarkNativeCallWithMapVariadicArray-8           	  200000	     10430 ns/op	    3806 B/op	      63 allocs/op
BenchmarkNativeCallWithFunction-8                   	  300000	      4255 ns/op	    1849 B/op	      26 allocs/op
BenchmarkNativeCallWithFunctionInt-8                	  300000	      4177 ns/op	    1849 B/op	      26 allocs/op
BenchmarkNativeCallWithFunctionString-8             	  300000	      4216 ns/op	    1849 B/op	      26 allocs/op
BenchmarkGoSliceQuickSort100000000-8                	  100000	     12020 ns/op	    3017 B/op	      99 allocs/op
BenchmarkGoSliceHeapSort100000000-8                 	  200000	      7030 ns/op	    2041 B/op	      60 allocs/op
BenchmarkJsArrayQuickSort10000-8                    	       1	2039089395 ns/op	428089408 B/op	 9507597 allocs/op
BenchmarkJsArrayMergeSort10000-8                    	       1	2236730446 ns/op	506091888 B/op	11524978 allocs/op
BenchmarkJsArrayHeapSort10000-8                     	       3	 456500062 ns/op	99366778 B/op	 2146712 allocs/op
BenchmarkCryptoAES-8                                	       1	1699194506 ns/op	317282080 B/op	 9414193 allocs/op
BenchmarkJSON_parse-8                               	   20000	     92281 ns/op	   26863 B/op	     605 allocs/op
BenchmarkNew-8                                      	   10000	    178183 ns/op	  149117 B/op	    1669 allocs/op
BenchmarkClone-8                                    	   10000	    238669 ns/op	  157207 B/op	    1467 allocs/op
BenchmarkString_match-8                             	   50000	     23128 ns/op	   44715 B/op	      92 allocs/op
BenchmarkString_replace-8                           	   50000	     24673 ns/op	   44475 B/op	      91 allocs/op
BenchmarkString_search-8                            	  100000	     20115 ns/op	   43185 B/op	      73 allocs/op
BenchmarkString_splitWithString-8                   	  100000	     17710 ns/op	    6417 B/op	     105 allocs/op
BenchmarkString_splitWithRegex-8                    	   50000	     37218 ns/op	   48965 B/op	     166 allocs/op
